rust-symcrypt/README.md

```diff
@@ -43,13 +43,10 @@ HKDF:
 - HmacSha384
 - HmacSha512
 
-GCM:
-- Encryption ( in place )
-- Decryption ( in place )
-
-ChaCha:
-- Encryption ( in place )
-- Decryption ( in place )
+Encryption: 
+- Gcm Encrypt/Decrypt

```

Comment:
NP: **AES**-GCM - there are GCM + other block ciphers defined (i.e. SM4-GCM)

Reply:
Interestingly, the underlying SymCrypt functions expose this as a generic GCM block cipher mode, but it does look like only AES is (currently) supported. I agree that it's probably best to message it as just AES-GCM for now. But I wonder if it will be confusing that the Rust primitive refers to just GCM? In theory, it could be used with other ciphers in the future.

Reply:
Right - in SymCrypt API GCM was written as a mode over a block cipher - but in practice we only support AES-GCM (https://github.com/microsoft/SymCrypt/blob/907622c6658877014ea3fea336e2efa407864fdc/inc/symcrypt_internal.h#L1584-L1587)

Reply:
For the GCM code we have `pub fn new(key: &[u8], cipher: BlockCipherType) -> Result<Self, SymCryptError> {` where it takes in `BlockCipherType` 

I think that if AES is only supported and will only be supported in the near future, it would make sense to change GCM  to AES-GCM similar to how we have AES-CBC right now. If you all agree, I will mark it as a Todo and file it under our (growing) backlog

Reply:
That's fine with me. I think in the future it makes sense for symcrypt-sys to expose this the same way SymCrypt does, i.e. generic block cipher modes that can be used with any supported block cipher. I'm okay with making them AES-specific for now, but we should also be consistent between CBC and GCM.

Reply:
Just realized that the way this is implemented isn't consistent in SymCrypt either - ECB, CBC, and CTR have cipher-specific functions whereas GCM is generic. So that makes my suggestion a bit hard to do. Let's circle back on this later.

Reply:
So for AES-KW(P) I am just exposing AES-KW(P) functionality directly.
This is similar to our existing AES-XTS functionality, where we only have an AES-specific flavor exposed.

ECB, CBC, and CtrMsb64 support both generic blockcipher and specific AES flavors.

GCM only supports the generic blockcipher, with the AES-GCM specific functionality being internal to SymCrypt.


I think I would argue for Rust-SymCrypt having only specific functions in the friendly layer, i.e. only support a subset of AES modes - do not support 3DES etc.
If we need to support some other block cipher in future we add new functions, not new enumeration values to a generic block cipher interface.

Reply:
But yeah, only exposing specifically "AES-GCM" functions from Rust-SymCrypt can be a backlog item.
We should really make this change (expose AES-GCM specific functions) in the SymCrypt layer as well - the current situation in the SymCrypt API is confused.